Functional Tab Widget

etruong opened the pull request:
- Fully functioning tab widget with no bugs (so far.. or so I've seen)
- I still need to refactor and reorganize the code as well as add comments, but at this point, I'm basically done with the functionality aspect

etruong committed:
complete to do list widget

etruong committed:
add tabs and styling

etruong committed:
implement correctly functioning tab view widget

kwokwilliam commented on package.json:
```diff
@@ -22,6 +22,7 @@
         "react": "^16.8.6",
         "react-dom": "^16.8.6",
         "react-flexbox-grid": "^2.1.2",
+        "react-markdown": "^4.1.0",
```
How much did it say the build size was before/after react-markdown was added?

etruong replied:
I'm honestly not sure how to check this, it didn't take long to add tho.

kwokwilliam commented on src/components/ContentMapping/ContentMapping.ts:
```diff
@@ -5,6 +5,10 @@ import { Stub, StubEditor } from '../ContentWidgets/_Stub/Stub';
 import { ContentSingularData } from '../_data/ContentSingularData';
 import { ExampleImage } from '../ContentWidgets/ExampleImage/ExampleImage';
 import { ExampleImageEditor } from '../ContentWidgets/ExampleImage/ExampleImageEditor';
+import { ToDoList } from '../ContentWidgets/ToDoList/ToDoList';
```
Remove todo list from this. That was just an example widget

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.css:
```diff
@@ -0,0 +1,92 @@
+    color: #ffffff;
+}
+
+.tab:hover {
```
`.tab` might be a pretty common class name, I would avoid using that. Maybe `tab-view` or something

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.css:
```diff
@@ -0,0 +1,92 @@
+    padding: 1rem;
+}
+
+.current-tab-content-editor > textarea, .current-tab-content > ReactMarkdown {
```
huh does this actually work? I would probably use an id here instead or a class `.react-markdown`.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.css:
```diff
@@ -0,0 +1,92 @@
+
+.add-tab-btn {
+    padding: 5px;
+    color: #ffffff;
```
It might be worth adding this to a global color variable that you can access. I don't remember if it's in this specific project but I had a css file in my index file for a project where I set common variables I can access through any CSS file, so maybe a variable white or so. 

ie

```    
:root {
    --theme-color: #005696;
}
```

I realize I haven't done this, but it might be good to start doing that so we can remain consistent.

kwokwilliam replied:
Can you add to the `index.css` file or `App.css` file that selector and set a `--white` to be `#ffffff`? You will be able to access it by using `var(--white)`

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.tsx:
```diff
@@ -0,0 +1,37 @@
+export const TabView: React.FC<ContentSingularData> = ({
+    tabView_content
+}) => {
+    
```
remove empty line

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.tsx:
```diff
@@ -0,0 +1,37 @@
+        return <></>
+    }
+    
+    let tabs = tabView_content!!.map((tab, index) => {
```
It would be preferable to have this within the ul tag below instead

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.tsx:
```diff
@@ -0,0 +1,37 @@
+    
+    let tabs = tabView_content!!.map((tab, index) => {
+        let tabClassName = "tab";
+        if (index == currentTab) {
```
change this to ternary within the curly braces for className:

```
className={`${index === currentTab ? "current-tab" : ""} tab`}
```

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+    editedContent,
+    setEditedContentOnChange
+}) => {
+
```
remove blank line

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+    const [currentTab, setcurrentTab] = useState(0);
+    const [renderView, setRenderView] = useState(false);
+
+    let copyTabViewContent: Tab[] = [];
```
Change this to `copiedTabViewContent`, and also use a ternary instead

```tsx
let copiedTabViewContent: Tab[] = editedContent.tabView_content ? [...editedContent.tabView_content!!] : []
```

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+        setEditedContentOnChange("tabView_content", copyTabViewContent);
+    }
+
+    let addNewTab = () => {
```
Make this function a const.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+        setEditedContentOnChange("tabView_content", copyTabViewContent);
+    }
+
+    let changeView = (event: React.MouseEvent<HTMLButtonElement>) => {
```
Make this function a const

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    let inputTabs = copyTabViewContent.map((tab, index) => {
+
```
remove empty line

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+    setEditedContentOnChange
+}) => {
+
+    const [currentTab, setcurrentTab] = useState(0);
```
capitalize C in `setcurrentTab`

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+            setEditedContentOnChange("tabView_content", copyTabViewContent);
+        }
+
+        let moveTab = (startEndTab: number, nextIndex: number) => {
```
all these functions should be const instead.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabViewEditor.tsx:
```diff
@@ -0,0 +1,105 @@
+
+    let currentTabContent = null;
+    let tabViewClass = "current-tab-content-editor";
+    if (copyTabViewContent.length > 0) {
```
I generally always prefer to have this render logic be within the return statement instead of using if statements. I think it is much more cleaner as well. You should do so here.

etruong committed:
resolve git pull requests

etruong committed:
merge with master

etruong committed:
add tests to TabView widget

etruong commented:
Hi, so I addressed the PR comments you made as well as, added additional test for the widget. Although I had trouble simulating an on change event to test the input, I push the tests I have so far that are working.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.css:
```diff
@@ -0,0 +1,92 @@
+}
+
+.tab-content-render-btn {
+    background-color: #420DAB;
```
If no global variables for this exist, can you create one in the `index.css` file maybe, and do something like:

```css
:root {
  --main-color: #DAB420;
  --secondary-color: #420DAB;
  --white: #FFF;
}
```

kwokwilliam replied:
Then you would use it here by doing `background-color: var(--secondary-color);`

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.test.tsx:
```diff
@@ -0,0 +1,159 @@
+
+describe("TabView component", () => {
+
+    const fakeProps = [
```
Careful with this. I don't know if `const` makes everything immutable inside. I would instead just make this a `let fakeProps` and then inside `beforeEach`, reset it to equal this. That way you are confident that before each test the fakeProps are consistent.

kwokwilliam replied:
There are no easy ways to deep clone something (IE if you were to use `let b = [...fakeProps]`, `b[0]` would actually be a pointer to `fakeProps[0]` (I believe). Easiest way would be to use lodash's cloneDeep method. We should include that library as well.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.test.tsx:
```diff
@@ -0,0 +1,159 @@
+    const editedContent: ContentSingularData = { tabView_content : undefined };
+
+    // setEditedContentOnChange
+    const setEditedContentOnChange = jest.fn().mockImplementation((contentTitle: string, 
```
I don't know if you have to create a mock implementation for this. You may be able to just either have it return a value or spy on it to see if it is called with expected variables. This is because we should have a test for `setEditedContentOnChange` that expects valid values, and because it's sort of a dependency you expect it is called with certain values.

kwokwilliam commented on src/components/ContentWidgets/TabView/TabView.test.tsx:
```diff
@@ -0,0 +1,159 @@
+    });
+
+    it("removes tabs correctly", () => {
+        originalContent.tabView_content = [
```
You should make a const for this because repetition is unmaintainable.

kwokwilliam reviewed (approved):
LOL rip I never submitted this but I meant to. I'll make these changes so don't worry. They're just for my reference. I'll merge this branch.

kwokwilliam committed:
Merged master and added a comment, and also set all packages to use ~

kwokwilliam committed:
refactored css stuffs
